src/app/context/UserSessionContext.tsx

```diff
@@ -0,0 +1,115 @@
+"use client";
+import { IdentifiableUser } from "@interfaces/type";
+import {
+  browserLocalPersistence,
+  GoogleAuthProvider,
+  onAuthStateChanged,
+  setPersistence,
+  signInWithPopup,
+  signOut,
+} from "firebase/auth";
+import React from "react";
+import { auth } from "@project/firebase";
+import { addUser, getUser } from "@services/client/user";
+import { useRouter } from "next/navigation";
+
+interface Session {
+  isAuthenticated: boolean;
+  isLoading: boolean;
+  error: string | null;
+}
+
+interface IUserSessionContext {
+  user: IdentifiableUser | null;
+  session: Session;
+  onSignIn: () => void;
+  onSignOut: () => Promise<void>;
+}
+
+const UserSessionContext = React.createContext<IUserSessionContext>(
+  {} as IUserSessionContext
+);
+
+export const UserSessionContextProvider = ({
+  children,
+}: {
+  children: React.ReactNode;
+}) => {
+  const [user, setUser] = React.useState<IdentifiableUser | null>(null);
+  const [session, setSession] = React.useState<Session>({
+    isAuthenticated: false,
+    isLoading: true,
+    error: null,
+  });
+  const router = useRouter();
+
+  React.useEffect(() => {
+    const unsub = onAuthStateChanged(auth, async (firebaseUser) => {
+      if (firebaseUser) {
+        const currUser = await getUser(firebaseUser?.uid);
+        setUser(currUser);
+        setSession({ isAuthenticated: true, isLoading: false, error: null });
```

Comment:
I wanted to `router.push("private/course/...")` but this ended in a infinite routing loop. We can explore that soon